cluster-provision/gocli/cmd/provision_manager.go

```diff
@@ -0,0 +1,519 @@
+	if err != nil {
+		return nil, err
+	}
+
+	targets := []string{}
+	for _, dir := range directories {
+		targets = append(targets, filepath.Base(dir))
+	}
+
+	logrus.Debug("Targets: ", targets)
+	return targets, nil
+}
+
+func processGitNameStatusChanges(rulesDB map[string][]string, targets []string, tag string) (map[string]bool, error) {
+	cmdOutput, err := runCommand("git", []string{"diff", "--name-status", tag})
+	if err != nil {
+		return nil, err
+	}
+
+	return processChanges(rulesDB, targets, cmdOutput)
+}
+
+func processChanges(rulesDB map[string][]string, targets []string, changes string) (map[string]bool, error) {
+	targetToRebuild := make(map[string]bool)
+	for _, target := range targets {
+		targetToRebuild[target] = false
+	}
+
+	printSection("Changed files")
+
+	errorFound := false
+	files := strings.Split(changes, "\n")
+	for _, nameStatus := range files {
+		if nameStatus == "" {
+			break
+		}
+
+		tokens := strings.Split(nameStatus, "\t")
+		if len(tokens) < 2 {
+			return nil, fmt.Errorf("wrong input syntax, should be <status>\\t<filename>")
+		}
+
+		status := tokens[0]
+		fileName := tokens[1]
+
+		if strings.HasPrefix(status, FILE_RENAMED) {
+			if len(tokens) != 3 {
+				return nil, fmt.Errorf("wrong input syntax, should be <status>\\t<old_filename>\\t<new_filename>")
+			}
+			fileName = tokens[2]
+		}
+
+		// Skip markdown files
+		if strings.HasSuffix(fileName, ".md") {
+			continue
+		}
+
+		match, err := matcher(rulesDB, fileName, status)
+		if err != nil {
+			errorFound = true
+			logrus.Error(err)
+			continue
+		}
+
+		if !errorFound {
+			logrus.Debug(status + " : " + fileName + " - [" + strings.Join(match, " ") + "]")
+		}
+
+		for _, target := range match {
+			if target != TARGET_NONE {
+				targetToRebuild[target] = true
+			}
+		}
+	}
+
+	if errorFound {
+		return nil, fmt.Errorf("Errors detected: files dont have a matching rule")
+	}
+
+	return targetToRebuild, nil
+}
+
+func matcher(rulesDB map[string][]string, fileName string, status string) ([]string, error) {
+	match, ok := rulesDB[fileName]
+	if ok {
+		return match, nil
+	}
+
+	match, ok = rulesDB[filepath.Dir(fileName)]
+	if ok {
+		return match, nil
+	}
+
+	candid := fileName
+	for candid != "." && candid != "/" {
+		candid = filepath.Dir(candid)
+		match, ok = rulesDB[candid+"/*"]
+		if ok {
+			return match, nil
+		}
+	}
+
+	if status != FILE_DELETED {
+		return nil, fmt.Errorf("Failed to find a rule for " + fileName)
+	}
+
+	return nil, nil
+}
+
+func getKubevirtciTag() (string, error) {
+	const STATUS_OK = 200
+	const kubevirtciTagUrl = "https://storage.googleapis.com/kubevirt-prow/release/kubevirt/kubevirtci/latest?ignoreCache=1"
+
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	resp, err := http.Get(kubevirtciTagUrl)
+	if err != nil {
+		return "", fmt.Errorf("ERROR: getting latest kubevirtci tag failed, error: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != STATUS_OK {
```

Comment:
Is this required? there is a check for errors on L285 already?

And there is a check for parsing the body below too?

Reply:
This why it is more explicit, there could be other errors like 404 or so, letting the previous checks pass.

Once you got status 200 you know there is content, you use that content as git hash
The content is valid basically, there is no need to check what is the content.
If theoretically the content is bad the git command will fail, but again, never saw it happens
and once we check status 200 it is to see that the server said all good and we can expect output and a valid one